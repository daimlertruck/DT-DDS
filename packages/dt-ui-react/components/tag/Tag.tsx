import { BaseProps } from '@dt-ui/react-core';

import { CloseIcon } from '../../core/assets';

import { TagBorder, TagColor, TagSize, TagVariant } from './constants';
import { TagStyled, TagGroupStyled, TagButtonCloseStyled } from './Tag.styled';

type TagConditionalProps =
  | {
      onClick?: (event: React.MouseEvent<HTMLDivElement>) => void;
      onClose?: never;
    }
  | {
      onClose?: (event: React.MouseEvent<HTMLButtonElement>) => void;
      onClick?: never;
    };

export type TagProps = {
  variant?: TagVariant;
  color?: TagColor;
  border?: TagBorder;
  size?: TagSize;
  isDisabled?: boolean;
  isClickable?: boolean;
} & TagConditionalProps &
  BaseProps;

export const Tag = ({
  style,
  children,
  dataTestId,
  variant = 'solid',
  color = 'primary',
  border = 'squared',
  size = 'small',
  isClickable = false,
  isDisabled = false,
  onClick,
  onClose,
}: TagProps) => {
  const dataTestIdName = dataTestId ?? 'tag';
  const isDismissible = !!onClose;
  const hasHoverStyle =
    !isDisabled && (!!onClick || isClickable || isDismissible);
  const hasClickableStyle = !isDismissible && hasHoverStyle;

  const handleClickTag = (event: React.MouseEvent<HTMLDivElement>) => {
    if (isDisabled || isDismissible) {
      return;
    }

    onClick?.(event);
  };

  return (
    <TagStyled
      aria-disabled={isDisabled}
      border={border}
      color={color}
      data-testid={dataTestIdName}
      hasHover={hasHoverStyle}
      isClickable={hasClickableStyle}
      isDisabled={isDisabled}
      isDismissible={isDismissible}
      onClick={handleClickTag}
      size={size}
      style={style}
      variant={variant}
    >
      {children}

      {isDismissible ? (
        <TagButtonCloseStyled
          aria-disabled={isDisabled}
          color={color}
          data-testid={`${dataTestIdName}-close-button`}
          disabled={isDisabled}
          hasHover={hasHoverStyle}
          onClick={onClose}
          variant={variant}
        >
          <CloseIcon height={12} width={12} />
        </TagButtonCloseStyled>
      ) : null}
    </TagStyled>
  );
};

export const TagGroup = ({ children, style }: BaseProps) => {
  return <TagGroupStyled style={style}>{children}</TagGroupStyled>;
};
