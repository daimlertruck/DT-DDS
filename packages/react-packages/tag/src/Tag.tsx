import { BaseProps } from '@dt-ui/react-core';
import { Tooltip, TooltipBackground } from '@dt-ui/react-tooltip';
import { forwardRef, useEffect, useRef, useState } from 'react';

import { CloseIcon } from '../../../dt-ui-react/core/assets';

import { TagBorder, TagColor, TagSize, TagVariant } from './constants';
import { TagStyled, TagButtonCloseStyled } from './Tag.styled';

export type TagProps = {
  variant?: TagVariant;
  color?: TagColor;
  border?: TagBorder;
  size?: TagSize;
  isDisabled?: boolean;
  isClickable?: boolean;
  onClick?: (
    event:
      | React.MouseEvent<HTMLDivElement>
      | React.KeyboardEvent<HTMLDivElement>
  ) => void;
  onClose?: (
    event:
      | React.MouseEvent<HTMLButtonElement>
      | React.KeyboardEvent<HTMLButtonElement>
  ) => void;
  tooltipVariant?: TooltipBackground;
} & BaseProps;

export const Tag = forwardRef<HTMLDivElement, TagProps>(
  (
    {
      style,
      children,
      dataTestId,
      variant = 'solid',
      color = 'primary',
      border = 'squared',
      size = 'small',
      isClickable = false,
      isDisabled = false,
      onClick,
      onClose,
      tooltipVariant = 'opacity',
    },
    ref
  ) => {
    const dataTestIdName = dataTestId ?? 'tag';
    const isDismissible = !!onClose;
    const hasHoverStyle =
      !isDisabled && (!!onClick || isClickable || isDismissible);
    const hasClickableStyle = !isDismissible && hasHoverStyle;

    const handleClickTag = (
      event:
        | React.MouseEvent<HTMLDivElement>
        | React.KeyboardEvent<HTMLDivElement>
    ) => {
      if (isDisabled || isDismissible) {
        return;
      }

      onClick?.(event);
    };

    const cellRef = useRef<HTMLDivElement>(null);
    const [isOverflow, setIsOverflow] = useState(false);

    const handleClickOnEnterKey = (
      event: React.KeyboardEvent<HTMLDivElement>
    ) => {
      if (event.code === 'Enter') {
        handleClickTag(event);
      }
    };

    const handleCloseOnEnterKey = (
      event: React.KeyboardEvent<HTMLButtonElement>
    ) => {
      if (event.code === 'Enter') {
        onClose?.(event);
      }
    };

    useEffect(() => {
      if (cellRef.current) {
        const { current } = cellRef;

        setIsOverflow(current.scrollWidth > current.clientWidth);
      }
    }, [children]);

    return (
      <Tooltip>
        <TagStyled
          aria-disabled={isDisabled}
          border={border}
          color={color}
          data-testid={dataTestIdName}
          hasHover={hasHoverStyle}
          isClickable={hasClickableStyle}
          isDisabled={isDisabled}
          isDismissible={isDismissible}
          onClick={handleClickTag}
          {...(isClickable && { onKeyDown: handleClickOnEnterKey })}
          ref={ref}
          size={size}
          style={style}
          {...(isClickable && { tabIndex: 0 })}
          variant={variant}
        >
          <span ref={cellRef}>{children}</span>

          {isDismissible ? (
            <TagButtonCloseStyled
              aria-disabled={isDisabled}
              color={color}
              data-testid={`${dataTestIdName}-close-button`}
              disabled={isDisabled}
              onClick={onClose}
              onKeyDown={handleCloseOnEnterKey}
              tabIndex={0}
            >
              <CloseIcon height={12} width={12} />
            </TagButtonCloseStyled>
          ) : null}
        </TagStyled>
        <Tooltip.Content background={tooltipVariant}>
          {isOverflow ? children : ''}
        </Tooltip.Content>
      </Tooltip>
    );
  }
);
