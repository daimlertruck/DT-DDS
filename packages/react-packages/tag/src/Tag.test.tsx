import { withProviders } from '@dt-ui/react-core';
import { fireEvent, render } from '@testing-library/react';

import { Tag, TagProps } from './Tag';

describe('<Tag /> component', () => {
  const ProvidedTag = withProviders(Tag);

  it.each`
    isDisabled | expectedClickTimes
    ${true}    | ${0}
    ${false}   | ${1}
  `(
    'should trigger the onClick function $expectedClickTimes times if isDisabled is $isDisabled, and close button should not exist if onClose is undefined',
    ({ isDisabled, expectedClickTimes }) => {
      const onClick = jest.fn();

      const { getByTestId, queryByTestId } = render(
        <ProvidedTag dataTestId='tag' isDisabled={isDisabled} onClick={onClick}>
          Click me
        </ProvidedTag>
      );

      fireEvent.click(getByTestId('tag'));

      expect(onClick).toHaveBeenCalledTimes(expectedClickTimes);
      expect(queryByTestId('tag-close-button')).toBeNull();
    }
  );

  it.each`
    isDisabled | expectedCloseTimes
    ${true}    | ${0}
    ${false}   | ${1}
  `(
    'should trigger the close button $expectedCloseTimes times if isDisabled is $isDisabled and trigger onClose but not onClick',
    ({ isDisabled, expectedCloseTimes }) => {
      const onClick = jest.fn();
      const onClose = jest.fn();

      const { getByTestId } = render(
        <ProvidedTag
          dataTestId='tag'
          isDisabled={isDisabled}
          onClick={onClick}
          onClose={onClose}
        >
          Click me
        </ProvidedTag>
      );

      fireEvent.click(getByTestId('tag'));
      fireEvent.click(getByTestId('tag-close-button'));

      expect(onClick).toHaveBeenCalledTimes(0);
      expect(onClose).toHaveBeenCalledTimes(expectedCloseTimes);
    }
  );

  it.each`
    variant       | color            | border       | size        | isClickable | isDisabled | onClick      | onClose
    ${'solid'}    | ${'primary'}     | ${'rounded'} | ${'medium'} | ${false}    | ${false}   | ${undefined} | ${undefined}
    ${'outlined'} | ${'secondary'}   | ${'squared'} | ${'small'}  | ${true}     | ${false}   | ${undefined} | ${undefined}
    ${'solid'}    | ${'informative'} | ${'rounded'} | ${'medium'} | ${false}    | ${false}   | ${jest.fn()} | ${undefined}
    ${'solid'}    | ${'warning'}     | ${'rounded'} | ${'medium'} | ${false}    | ${false}   | ${undefined} | ${jest.fn()}
    ${'solid'}    | ${'error'}       | ${'rounded'} | ${'medium'} | ${false}    | ${true}    | ${jest.fn()} | ${jest.fn()}
  `(
    'should render the tag with variant $variant, color $color, border $border, size $size, isClickable as $isClickable, isDisabled as $isDisabled, onClick and onClose',
    ({
      variant,
      color,
      border,
      size,
      isClickable,
      isDisabled,
      onClick,
      onClose,
    }: TagProps) => {
      const { getByTestId } = render(
        <ProvidedTag
          border={border}
          color={color}
          dataTestId='tag'
          isClickable={isClickable}
          isDisabled={isDisabled}
          onClick={onClick}
          onClose={onClose}
          size={size}
          variant={variant}
        >
          Click me
        </ProvidedTag>
      );

      expect(getByTestId('tag')).toMatchSnapshot();
    }
  );
});
