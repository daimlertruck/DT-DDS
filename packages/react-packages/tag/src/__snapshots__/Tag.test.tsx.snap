// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tag /> component should render the tag with variant outlined, color secondary, border squared, size small, isClickable as true, isDisabled as false, onClick and onClose 1`] = `
.emotion-0 {
  font-size: 0.625rem;
  font-weight: 700;
  line-height: 0.875rem;
  padding: 2px 8px;
  height: 18px;
  border: 1px solid #3e3e3e;
  color: #3e3e3e;
  background-color: #c5c5c5;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-transform: uppercase;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: -webkit-max-content;
  max-width: -moz-max-content;
  max-width: max-content;
  gap: 4px;
  border-radius: 0px;
  width: 150px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
}

.emotion-0:hover {
  border-color: #1d1d1d;
  color: #1d1d1d;
  background-color: #696969;
}

.emotion-0>:first-child {
  white-space: nowrap;
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
}

.emotion-0:focus-visible {
  outline: 2px solid #1d1d1d;
  outline-offset: 1px;
}

<div
  aria-disabled="false"
  class="emotion-0 emotion-1"
  color="secondary"
  data-testid="tag"
  tabindex="0"
>
  <span>
    Click me
  </span>
</div>
`;

exports[`<Tag /> component should render the tag with variant solid, color error, border rounded, size medium, isClickable as false, isDisabled as true, onClick and onClose 1`] = `
.emotion-0 {
  font-size: 0.75rem;
  font-weight: 700;
  line-height: 1rem;
  padding: 8px 12px;
  height: 32px;
  border: 0px;
  color: #ffffff;
  background-color: #ffe6e6;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-transform: uppercase;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: -webkit-max-content;
  max-width: -moz-max-content;
  max-width: max-content;
  gap: 4px;
  border-radius: 32px;
  width: 150px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: not-allowed;
}

.emotion-0>:first-child {
  white-space: nowrap;
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
}

.emotion-0:focus-visible {
  outline: 2px solid #1d1d1d;
  outline-offset: 1px;
}

.emotion-2 {
  background-color: transparent;
  color: currentColor;
  display: inherit;
  border: 0;
  cursor: pointer;
}

.emotion-2:disabled {
  cursor: not-allowed;
}

.emotion-2:focus-visible {
  outline: 2px solid #1d1d1d;
  outline-offset: 1px;
}

<div
  aria-disabled="true"
  class="emotion-0 emotion-1"
  color="error"
  data-testid="tag"
>
  <span>
    Click me
  </span>
  <button
    aria-disabled="true"
    class="emotion-2 emotion-3"
    color="error"
    data-testid="tag-close-button"
    disabled=""
    tabindex="0"
  >
    <svg
      fill="none"
      height="12"
      viewBox="0 0 24 24"
      width="12"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19 6.41L17.59 5L12 10.59L6.41 5L5 6.41L10.59 12L5 17.59L6.41 19L12 13.41L17.59 19L19 17.59L13.41 12L19 6.41Z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;

exports[`<Tag /> component should render the tag with variant solid, color informative, border rounded, size medium, isClickable as false, isDisabled as false, onClick and onClose 1`] = `
.emotion-0 {
  font-size: 0.75rem;
  font-weight: 700;
  line-height: 1rem;
  padding: 8px 12px;
  height: 32px;
  border: 0px;
  color: #ffffff;
  background-color: #00677f;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-transform: uppercase;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: -webkit-max-content;
  max-width: -moz-max-content;
  max-width: max-content;
  gap: 4px;
  border-radius: 32px;
  width: 150px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
}

.emotion-0:hover {
  background-color: #003946;
  color: #ffffff;
}

.emotion-0>:first-child {
  white-space: nowrap;
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
}

.emotion-0:focus-visible {
  outline: 2px solid #1d1d1d;
  outline-offset: 1px;
}

<div
  aria-disabled="false"
  class="emotion-0 emotion-1"
  color="informative"
  data-testid="tag"
>
  <span>
    Click me
  </span>
</div>
`;

exports[`<Tag /> component should render the tag with variant solid, color primary, border rounded, size medium, isClickable as false, isDisabled as false, onClick and onClose 1`] = `
.emotion-0 {
  font-size: 0.75rem;
  font-weight: 700;
  line-height: 1rem;
  padding: 8px 12px;
  height: 32px;
  border: 0px;
  color: #ffffff;
  background-color: #00677f;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-transform: uppercase;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: -webkit-max-content;
  max-width: -moz-max-content;
  max-width: max-content;
  gap: 4px;
  border-radius: 32px;
  width: 150px;
}

.emotion-0>:first-child {
  white-space: nowrap;
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
}

.emotion-0:focus-visible {
  outline: 2px solid #1d1d1d;
  outline-offset: 1px;
}

<div
  aria-disabled="false"
  class="emotion-0 emotion-1"
  color="primary"
  data-testid="tag"
>
  <span>
    Click me
  </span>
</div>
`;

exports[`<Tag /> component should render the tag with variant solid, color warning, border rounded, size medium, isClickable as false, isDisabled as false, onClick and onClose 1`] = `
.emotion-0 {
  font-size: 0.75rem;
  font-weight: 700;
  line-height: 1rem;
  padding: 8px 12px;
  height: 32px;
  border: 0px;
  color: #ffffff;
  background-color: #fab700;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-transform: uppercase;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: -webkit-max-content;
  max-width: -moz-max-content;
  max-width: max-content;
  gap: 4px;
  border-radius: 32px;
  width: 150px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-0:hover {
  background-color: #8a6500;
  color: #ffffff;
}

.emotion-0>:first-child {
  white-space: nowrap;
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
}

.emotion-0:focus-visible {
  outline: 2px solid #1d1d1d;
  outline-offset: 1px;
}

.emotion-2 {
  background-color: transparent;
  color: currentColor;
  display: inherit;
  border: 0;
  cursor: pointer;
}

.emotion-2:disabled {
  cursor: not-allowed;
}

.emotion-2:focus-visible {
  outline: 2px solid #1d1d1d;
  outline-offset: 1px;
}

<div
  aria-disabled="false"
  class="emotion-0 emotion-1"
  color="warning"
  data-testid="tag"
>
  <span>
    Click me
  </span>
  <button
    aria-disabled="false"
    class="emotion-2 emotion-3"
    color="warning"
    data-testid="tag-close-button"
    tabindex="0"
  >
    <svg
      fill="none"
      height="12"
      viewBox="0 0 24 24"
      width="12"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19 6.41L17.59 5L12 10.59L6.41 5L5 6.41L10.59 12L5 17.59L6.41 19L12 13.41L17.59 19L19 17.59L13.41 12L19 6.41Z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;
